### ISSUE TITLE
Generalizing to Vectors and Matrices

### ISSUE THREAD

**Issue by mratsim** [mratsim/nim-rmad#4]
To generalize to matrices and Vector, the `BackProp` type must be changed to support `T -> U` functions, T and U being `SomeReal, Matrix32, DMatrix32, Vector32, DVector32` from `linalg`

This works but is not enough:

```Nim
import linalg

type
  BackProp[T, U] = proc (gradient: T): U {.noSideEffect.}
    ## Proc for Backward propagation are typed `BackProp` and are (implicit) closures without side-effects
    # To ease search, backward propagation procedures are prefixed with bp_

  Node[T, U] = object
    ## Represent an operation
    ## Stores the gradient transformation for backprop in weights
    ## Stores indices of parent operation in parents
    weights: array[2, BackProp[T, U]]
    parents: array[2, int] #ref indices to parent nodes

  Context*[T, U] = object
    ## Tape / Wengert list. Contains the list of applied operations
    nodes: ref seq[Node[T, U]]

# Templates in Nim are always inlined. They are used for performance reason to save on function calls costs.

proc newContext*[T, U]: Context[T, U] {.noSideEffect.} =
  ## Initialize a context (Tape / Wengert list)
  result.nodes = new seq[Node[T, U]]
  result.nodes[] = @[]

let ctx = newContext[float32, Matrix32[2,2]]
```

The main issue is that now all proc must transform a scalar to a 2x2 Matrix.
We need at minimum:
- Scalar to Vector (backprop of Softmax)
- Vector to Matrix (backprop of Flatten)
- Matrix NxN to Matrix MxM (Convolutional Layer and MaxPool Layer)

What was tried:
```Nim
import linalg

type

  Tensor[float32] = float32 or DVector32

  BackProp[T] = proc (gradient: Tensor[T]): Tensor[T] {.noSideEffect.}
    ## Proc for Backward propagation are typed `BackProp` and are (implicit) closures without side-effects
    # To ease search, backward propagation procedures are prefixed with bp_

  Node[T] = object
    ## Represent an operation
    ## Stores the gradient transformation for backprop in weights
    ## Stores indices of parent operation in parents
    weights: array[2, BackProp[T]]
    parents: array[2, int] #ref indices to parent nodes

  Context*[T] = object
    ## Tape / Wengert list. Contains the list of applied operations
    nodes: ref seq[Node[T]]

# Templates in Nim are always inlined. They are used for performance reason to save on function calls costs.

proc newContext*(T: typedesc[SomeReal]): Context[T] {.noSideEffect.} =
  ## Initialize a context (Tape / Wengert list)
  result.nodes = new seq[Node[T]]
  result.nodes[] = @[]

let ctx = newContext(float32)

# Cannot evaluate at compile time: T
```

```Nim
import linalg

type

  Tensor = float32 or DVector32

  BackProp = proc (gradient: Tensor): Tensor {.noSideEffect.}
    ## Proc for Backward propagation are typed `BackProp` and are (implicit) closures without side-effects
    # To ease search, backward propagation procedures are prefixed with bp_

  Node = object
    ## Represent an operation
    ## Stores the gradient transformation for backprop in weights
    ## Stores indices of parent operation in parents
    weights: array[2, BackProp]
    parents: array[2, int] #ref indices to parent nodes

  Context* = object
    ## Tape / Wengert list. Contains the list of applied operations
    nodes: ref seq[Node]

# Templates in Nim are always inlined. They are used for performance reason to save on function calls costs.

proc newContext*(T: typedesc[SomeReal]): Context {.noSideEffect.} =
  ## Initialize a context (Tape / Wengert list)
  result.nodes = new seq[Node]
  result.nodes[] = @[]

let ctx = newContext(float32)
# Error: 'BackProp' is not a concrete type.
```

There is also one variation of the Tensor ... = float32 or DVector32 ... with `Cannot instantiate error` with this kind of declaration if I remember properly:

`BackProp[T: Tensor, U: Tensor] = proc (gradient: T): U {.noSideEffect.}`

**Comment by mratsim** [comment]
Forum post: [https://forum.nim-lang.org/t/2907](https://forum.nim-lang.org/t/2907)

Using this object variant solutions works:
```Nim
import linalg

type
  ## Unfortunately Nim typechecker won't let us store BackProp[float32, Vector32]
  ## with BackProp[Vector32, Matrix32] in the same list so I have to store each type
  ## of BackProp[T, U] in it's own list
  BackPropKind = enum
    ## List of BackProp[T,U] types
    bpFF,
    bpFV,
    bpVF,
    bpVV,
    bpFM,
    bpVM,
    bpMF,
    bpMV,
    bpMM

  BackProp = ref BackPropObj
  BackPropObj = object
    ## Object storing the procs for Backward propagation are typed `BackProp` and are (implicit) closures without side-effects
    # To ease search, backward propagation procedures are prefixed with bp_
    case kind: BackPropKind
    of bpFF:
        ff: proc (gradient: float32): float32 {.noSideEffect.}
    of bpFV:
        fv: proc (gradient: float32): DVector32 {.noSideEffect.}
    of bpVF:
        vf: proc (gradient: DVector32): float32 {.noSideEffect.}
    of bpVV:
        vv: proc (gradient: DVector32): DVector32 {.noSideEffect.}
    of bpFM:
        fm: proc (gradient: float32): DMatrix32 {.noSideEffect.}
    of bpVM:
        vm: proc (gradient: DVector32): DMatrix32 {.noSideEffect.}
    of bpMF:
        mf: proc (gradient: DMatrix32): float32 {.noSideEffect.}
    of bpMV:
        mv: proc (gradient: DMatrix32): DVector32 {.noSideEffect.}
    of bpMM:
        mm: proc (gradient: DMatrix32): DMatrix32 {.noSideEffect.}

  Node = object
    ## Represent an operation
    ## Stores the gradient transformation for backprop in weights
    ## Stores indices of parent operation in parents
    weights: array[2, BackProp]
    parents: array[2, int] #ref indices to parent nodes

  Context* = object
    ## Tape / Wengert list. Contains the list of applied operations
    nodes: ref seq[Node]

# Templates in Nim are always inlined. They are used for performance reason to save on function calls costs.

proc newContext*: Context {.noSideEffect.} =
  ## Initialize a context (Tape / Wengert list)
  result.nodes = new seq[Node]
  result.nodes[] = @[]

let ctx = newContext()
```

It's not possible to use the statically typed Matrix and Vector instead of DMatrix and DVector unfortunately, probably because of the Matrix32[M, N: static[int]] signature.

So we lose compiler autodetection for:
 - square matrices and call of an optimized BLAS procedure
 - incompatible shapes, for Vector-Matrices and Matrices-Matrices product

Bonus point to de-uglify this type section >_>

**Comment by mratsim** [comment]
Further attempts by @Varriount were unsuccessful, namely:

```Nim
import linalg


{.experimental.}

# Global templates
template foreach(expression, iter: untyped): untyped =
  for element in mitems(iter):
    expression(element)


type
  BackPropRoot = ref object of RootObj
  BackPropClosure[T, U] = proc (gradient: T): U {.noSideEffect, closure.}

  ## Proc for Backward propagation are typed `BackProp` and are (implicit) closures without side-effects
  ## To ease search, backward propagation procedures are prefixed with bp_
  BackProp[T, U] = ref object of BackPropRoot
    procedure: BackPropClosure[T, U]

  ## Represent an operation
  ## Stores the gradient transformation for backprop in weights
  ## Stores indices of parent operation in parents
  Node = object
    weights: array[2, BackPropRoot]
    parents: array[2, int] #ref indices to parent nodes

  ## Tape / Wengert list. Contains the list of applied operations
  Context* = object
    nodes: ref seq[Node]


# BackProp implementations
method call(target: BackPropRoot, param: int): int {.base.} =
  return param

method call[T, U](target: BackProp[T, U], param: T): U =
  result = target.p(param)

proc newBackProp[T, U](p: BackPropClosure): BackProp =
  new(result)
  result.procedure = p


# Context implementations
proc newContext*(): Context {.noSideEffect.} =
  ## Initialize a context (Tape / Wengert list)
  result.nodes = new seq[Node]


proc main() = 
  proc squareInt(gradient: int): int {.noSideEffect, closure.} =
    result = gradient * gradient

  proc squareFloat(gradient: float): float {.noSideEffect, closure.} =
    result = gradient * gradient


  var
    ctx = newContext()
    n = newNode()
```

And

```Nim
import linalg


{.experimental.}

# Global templates
template foreach(expression, iter: untyped): untyped =
  for element in mitems(iter):
    expression(element)


type
  BackPropRoot = ref object of RootObj
  BackPropClosure[T, U] = proc (gradient: T): U {.noSideEffect, closure.}

  ## Proc for Backward propagation are typed `BackProp` and are (implicit) closures without side-effects
  ## To ease search, backward propagation procedures are prefixed with bp_
  BackProp[T, U] = ref object of BackPropRoot
    procedure: BackPropClosure[T, U]

  ## Represent an operation
  ## Stores the gradient transformation for backprop in weights
  ## Stores indices of parent operation in parents
  Node = object
    weights: array[2, BackPropRoot]
    parents: array[2, int] #ref indices to parent nodes

  ## Tape / Wengert list. Contains the list of applied operations
  Context* = object
    nodes: ref seq[Node]


# BackProp implementations
method call[T, U](target: BackPropRoot, param: T): U {.base.} =
  # Raise exception here
  echo "Hi"
  return result

method call[T, U](target: BackProp[T, U], param: T): U =
  echo "Ho"
  result = target.procedure(param)

proc newBackProp[T, U](p: BackPropClosure[T, U]): BackProp[T, U] =
  new(result)
  result.procedure = p


# Context implementations
proc newContext*(): Context {.noSideEffect.} =
  ## Initialize a context (Tape / Wengert list)
  result.nodes = new seq[Node]


proc main() = 
  proc squareInt(gradient: int): int {.noSideEffect, closure.} =
    result = gradient * gradient

  proc squareFloat(gradient: float): float {.noSideEffect, closure.} =
    result = gradient * gradient


  var
    ctx = newContext()
    n: Node

  n.weights[0] = newBackProp(squareFloat)
  n.weights[1] = newBackProp(squareInt)
  call(n.weights[0], 1)
  call(n.weights[0], 1.0)
  call(n.weights[1], 1)
  call(n.weights[1], 1.0)

  ctx.nodes.add(n)
```

**Comment by mratsim** [comment]
Given the difficult to use `linalg` as is and future needs of 3D, 4D, 5D tensors for:

* height
* width
* batch size
* RGB color channels for color images
* depth for 3D or time for video

It is a must to develop a Nim ndarrays/tensor library to unify vectors and matrices and support at least 5 dimensions with GPU computation. Furthermore type signature between vector, matrices and more should be the same.

Work as started on the [Arraymancer project](https://github.com/mratsim/Arraymancer).

**Comment by mratsim** [comment]
Done in [Arraymancer/autograd](https://github.com/mratsim/Arraymancer/tree/master/src/autograd)
